### DISCUSSION TITLE
Extracting messages

### DISCUSSION THREAD

**Discussion by christopanayotovvia** [DoneDeal0/Talkr#14] [Ideas]
Iw would be helpful if there's a CLI to extract all messages and build JSON :

ex. talkt extract /**/*.ts|x/ should find all keys, sort them alphabetically, nest them if needed ( nesting would be one option, the other is to be dot-notated ) and return as json. Example results

```
{
 "name": {
  "first": 'First name'
  "last": 'Last name'
 }
}

{
"name.first": "First name",
"name.last": "Last name"
}
```

**Comment by DoneDeal0** [comment]
Hi Christoph, thanks for your message and your interest in Talkr! However, I don't really see the added value of your proposal. What would be the point to find all the keys used in your app and then output a json file? You are supposed to have a json in the first place to use the package anyways. Also, some keys may be dynamic, so we couldn't infer the full path in some cases. 

For exemple, the following script could return `day.monday` or `day.sunday`.

```js
function DisplayDay({day}){
return (
<h1>{ T(`date.${date}`) }</h1>
)
}
```

But maybe I didn't understand well your proposal. If so, feel free to send a pr or to correct me.

**Reply by christopanayotovvia** [reply]
It's a common practice in enterprise applications to automatically extract messages, upload them for translators ( ex. localise ) and automatically download updates/translated messages upon build. See https://formatjs.io/docs/getting-started/message-extraction/ for example.

**Reply by DoneDeal0** [reply]
Hi, usually devs simply add their keys to lokalize or similar service, and launch a script to download the updated json files. Then, pro translators can directly update the translations in the lokalize app. Finding all your translations - including dynamic ones - in a complex codebase, is certainly doable. Alas, I don't have enough time to do it. If Talkr was widely used, it would worth the effort, but not at this stage. However, feel free to send a pr.

**Reply by ablankenship10** [reply]
Talkr doesn't really work this way at least the way I use it, maybe there are other ways. 100% of all my strings are in my default language file (en.json). There are no default strings or anything within the codebase. There is nothing to extract from the codebase other than keys to the json locations where strings would be. If, for example, you wanted to get translations done you would just upload the JSON file and select the languages you want and I would presume translators send you back your translation files as basically a mirrored copy of the original JSON object with all the same keys but with other language text filled in.

Only tricky aspect maybe in the case where keys don't exist for other tenses or number of values (zero, one, few, many, etc) that would require the translator to add additional keys and probably wouldn't be aware of the Talkr required format for that. Not sure of the standards around this stuff.

**Reply by ablankenship10** [reply]
I think, however, a middle ground of what @christopanayotovvia is asking @DoneDeal0 would be you know how it shows a warning in console if a key doesn't exist in your JSON that you tried to use in your code? Maybe a command you could run that could extract all those and put them in your json file for you to fill in instead of creating manually. But the thing with that is those warnings only show on render of a component containing a call to T(), not the codebase as a whole. Also, anyone could rename the T() function whatever they want in each file so detecting it in the code files themselves rather than at runtime is still not trivial (or is trivial? I think either wat is correct.. English is weird). 

IMO its just easier to fill in your JSON file first and then use the key in your code after it already exists. That said, is there a way to detect discrepancies between JSON files, like if I update en.ts but not fr.ts, would there be an easy way to diff them and find the updates I need to make to the other file?

**Reply by DoneDeal0** [reply]
A command line to check differences between json files should be doable. But it's not a big feature for Talkr, and there are probably existing tools to do it anyways. Actual improvements could be:

- making Talkr work with React Native.
- adding tests 
- find a way to drive adoption for this lib lol (it's the lightest 18n package out there, we should aim for more downloads)

**Reply by samuelstroschein** [reply]
Hi everyone, 

I am building an ecosystem of localization solutions that seem to partially serve @christopanayotovvia needs and @ablankenship10 idea regarding a CLI that can "lint" translations (see https://github.com/inlang/inlang/discussions/108. )

Talkr could work with the [inlang project](https://github.com/inlang/inlang) through a [converter](https://inlang.dev/docs/glossary/converter). Is there interest from your side?

@DoneDeal0 
PS I went for the "push fluent as standard format" approach (we chatted on reddit). 



![vscode-extension-pattern-extraction](https://user-images.githubusercontent.com/35429197/154853819-b941979b-3982-4ccf-b7b0-39bafd5b433d.gif)
_The VS Code extension is one of the tools of the inlang project and makes extraction of text easy in addition to showing the messages inline._

**Reply by DoneDeal0** [reply]
Hi Samuel,

Yes I remember our chat on Reddit, I'm glad to see that `Inlang` has made such progresses! A CLI that checks all missing translations in json files would be very nice indeed. 

The thing is that despite all its recent improvements (autocomplete, code refactor for faster performance and conditional rendering, official website, etc.), `Talkr` doesn't have a big user base yet.  So I don't know whether it's worth it to spend a lot of time in coding such a feature with converters & whatnot.

Except if it looks easy/quick to implement, or if there is a crazy demand for that, I'll put this task on the back burner for now. Of course, any PR is welcome.

**Comment by christopanayotovvia** [comment]
Hi All,
Thanks for sharing this. We’re currently using formatJS with formatted messages extractor and automatically upload/download from localise. Do you want to do a meeting on what are some issues that we have been facing ( and did workarounds ) and how your project approaches this?

Best,
Chris
On 20 Feb 2022, 19:08 +0200, Samuel Stroschein ***@***.***>, wrote:
> Hi everyone,
> I am building an ecosystem of localization solutions that seem to partially serve @christopanayotovvia needs and @ablankenship10 idea regarding a CLI that can "lint" translations (see inlang/inlang#108. )
> Talkr could work with the inlang project through a converter. Is there interest from your side?
> @DoneDeal0
> PS I went for the "push fluent as standard format" approach (we chatted on reddit).
>
> The VS Code extension is one of the tools of the inlang project and makes extraction of text easy in addition to showing the messages inline.
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> Triage notifications on the go with GitHub Mobile for iOS or Android.
> You are receiving this because you were mentioned.Message ID: ***@***.***>

-- 
This communication and any attachments may contain confidential information 
and are intended to be viewed only by the intended recipients. If you have 
received this message in error, please notify the sender immediately by 
replying to the original message and then delete all copies of the email 
from your systems.

**Reply by samuelstroschein** [reply]
@christopanayotovvia 

Bumping in case you missed it. 

I am super curious about your workflows and workarounds. Can you write me a mail to `proxy` at `stroschein.me` ?
